src/upgrade/nc_upgrade_manager.js

```diff
@@ -0,0 +1,266 @@
+/* Copyright (C) 2016 NooBaa */
+"use strict";
+
+const os = require('os');
+const _ = require('lodash');
+const path = require('path');
+const pkg = require('../../package.json');
+const dbg = require('../util/debug_module')('UPGRADE');
+const { should_upgrade, load_required_scripts } = require('./upgrade_utils');
+
+dbg.set_process_name('Upgrade');
+const hostname = os.hostname();
+
+const CONFIG_DIR_LOCKED = 'CONFIG_DIR_LOCKED';
+const CONFIG_DIR_UNLOCKED = 'CONFIG_DIR_UNLOCKED';
+
+/////////////////////////////
+//       NC UPGRADES       //
+/////////////////////////////
+
+/**
+ * NC upgrades are Online(*) upgrades that should be executed as follows - 
+ * 1. Backup the config directory.
+ * 2. Iterate all hosts one by one - 
+ *    2.1. Upgrade the host's RPM
+ *    2.2. Restart noobaa service on the host.
+ * 3. Run `noobaa-cli upgrade start` - will initialize config directory upgrade
+ * 
+ * * Online upgrade - during the config directory upgrade, all creations/updates/deletions of entities like buckets/accounts will be blocked.
+ * 
+ * system.json upgrade information - 
+ * 1. RPM upgrades are per host, therefore they will be presented under the hostname object, in the upgrade history section.
+ * 2. Config Directory are cluster scoped - therefore, the config_directory key will be on the first level and in_progress_upgrade and upgrade_history of the config directory will be under it.
+ * Example -
+ * {
+ *   [hostname]: { current_version: 5.18.0, upgrade_history: {{ from_version: 5.17.0, to_version: 5.18.0, completed_scripts: [], timestamp }}}
+ *   config_directory: { config_dir_version: 1,
+ *                       upgrade_package_version: 5.17.0,
+ *                       phase: 'CONFIG_DIR_LOCKED',
+ *                       in_progress_upgrade: { from_version: 5.17.0, to_version: 5.18.0, config_dir_from_version: 0, config_dir_to_version: 1, completed_scripts: [], timestamp }},
+ *                       upgrade_history: {}
+ * }
+ */
+
+
+/**
+ * update_rpm_upgrade updates an RPM upgrade of a specific host in the system.json file if there was an upgrade
+ * an upgrade is detected when the version in system.json is lower than the version in package.json.
+ * @returns {Promise<Void>}
+ */
+async function update_rpm_upgrade(config_fs) {
+    const system_data = await config_fs.get_system_config_file({ silent_if_missing: true });
+    if (!system_data) return;
+
+    const from_version = system_data?.[hostname]?.current_version;
+    const to_version = pkg.version;
+
+    if (!should_upgrade(from_version, to_version)) return;
+
+    const this_upgrade = { timestamp: Date.now(), from_version, to_version };
+    system_data[hostname].current_version = to_version;
+    system_data[hostname]?.upgrade_history?.successful_upgrades.unshift(this_upgrade);
+    await config_fs.update_system_json_with_retries(JSON.stringify(system_data));
```

Comment:
[update issue](https://github.com/noobaa/noobaa-core/issues/8343) is relevant here too